README.md
# et_console_app-cookbook

A cookbook to provision an Apache server for the purpose of serving Console and it's accompanying Mobile Web app.

The Mobile Web app should be broken out into its own cookbook(s) if that becomes a separate property.

## Supported Platforms

* Ubuntu 12.04

## Attributes

Key                                    | Type            | Description                                      | Default
---                                    | ----            | -----------                                      | -------
`['apache']['listen_ports']`           | String or Array | Ports Apache listens on                          | `%w(8080)`
`['apache']['contact']`                | String          | Apache server contact                            | `'devops@evertrue.com'`
`['et_console_app']['deploy_to']`      | String          | Path to Console                                  | `'/var/www/console.evertrue.com'`
`['et_console_app']['docroot']`        | String          | Docroot for Console                              | `"#{node['et_console_app']['deploy_to']}/current/release"`
`['et_console_app']['server_name']`    | String          | Primary domain for Console                       | `"#{domain_prefix}console.evertrue.com"`
`['et_console_app']['server_aliases']` | Array           | Secondary domains for Console                    | `["#{domain_prefix}dashboard.evertrue.com"]`
`['et_web_app']['deploy_to']`          | String          | Path to Web                                      | `'/var/www/web.evertrue.com'`
`['et_web_app']['docroot']`            | String          | Docroot for Web                                  | `"#{node['et_web_app']['deploy_to']}/current/release"`
`['et_web_app']['server_name']`        | String          | Primary domain for Web                           | `"#{domain_prefix}web.evertrue.com"`
`['apache']['default_modules']`        | Array           | Loaded Apache modules                            |`%w(status alias auth_basic authn_file authz_default authz_groupfile authz_host authz_user dir env mime setenvif deflate expires headers rewrite)`
`['apache']['status_allow_list']`      | String          | Addresses allow to get server status             | `'127.0.0.1 localhost ip6-localhost'`
`['apache']['ext_status']`             | Boolean         | Whether or not to provide extended server status | `true`


## Usage

### et_console_app::default

Include `et_console_app` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[et_console_app::default]"
  ]
}
```

## Contributing

1. Fork the repository on Github
2. Create a named feature branch (i.e. `add-new-recipe`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request

## License and Authors

Author:: EverTrue, Inc. (<jeff@evertrue.com>)


CHANGELOG.md
et_console_app cookbook CHANGELOG
============================
This file is used to list changes made in each version of the et_console_app cookbook.

v2.1.0
------
- Use our own Berkshelf API server for internal cookbooks
- Fix ChefSpec tests to have Node install command properly stubbed
- Add `(stage-)app.evertrue.com` as a vhost
    + This cookbook is increasingly inaccurately named…

v2.0.4
------
- Add missing hyphen

v2.0.3
------
- Hack attributes to work with stage-newvpc environment

v2.0.2
------
- Set RewriteLogLevel back to 0 to prevent huge logs from being written

v2.0.1
------
- Fix OPTIONS requests for better HAProxy health checks

v2.0.1
------
- Fix hard-coded setting of listen port for `web_app` resources

v2.0.0
------
- Switch to configure Apache to listen on port 8080 (breaking change)

v1.0.1
------
- Fix RewriteRule for cachebusting timestamped assets

v1.0.0
------
- Add installation of Node.js, NPM, Bower, Grunt, Git
- Fix ChefSpec tests for group ownership of some Apache folders
- Specify platform for ChefSpec tests

v0.1.0 (2014-03-21)
-------------------
- Initial release of et_console_app


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'
require 'chefspec/server'

RSpec.configure do |config|
  config.log_level = :fatal
end

ChefSpec::Server.create_client('et_console_app_spec', admin: true)


spec/recipes/default_spec.rb
require 'spec_helper'

describe 'et_console_app::default' do
  let(:chef_run) do
    ChefSpec::Runner.new(platform: 'ubuntu', version: '12.04') do |node|
      node.set['apache']['root_group'] = 'root'
      node.set['apache']['group'] = 'www-data'
      node.set['apache']['dir'] = '/etc/apache2'
    end.converge(described_recipe)
  end

  before do
    stub_command('[ -f /usr/local/share/node-version ] && [ "$(git rev-parse HEAD)" = "$(cat /usr/local/share/node-version)" ]')
      .and_return(true)
  end

  %w(
    apt::default
    node::default
    et_users::evertrue
    apache2::default
  ).each do |recipe|
    it "includes #{recipe}" do
      expect(chef_run).to include_recipe recipe
    end
  end

  it 'installs git' do
    expect(chef_run).to install_package 'git'
  end

  it 'creates /etc/apache2/conf.d/h5bp.conf' do
    expect(chef_run).to create_cookbook_file_if_missing('/etc/apache2/conf.d/h5bp.conf').with(
      user: 'root',
      group: 'root',
      mode: '0644'
    )
  end

  %w(
    /var/www/console.evertrue.com
    /var/www/web.evertrue.com
    /var/www/app.evertrue.com
  ).each do |path|
    it "creates directory #{path}" do
      expect(chef_run).to create_directory(path).with(
        user: 'deploy',
        group: 'www-data',
        mode: '2775'
      )
    end
  end
end


test/integration/default/serverspec/default_spec.rb
require 'spec_helper'

describe 'Apache Service' do
  %w(8080).each do |port|
    describe port(port) do
      it { should be_listening }
    end
  end

  describe service('apache2') do
    it { should be_running }
    it { should be_enabled }
  end
end

describe 'Apache Virtual Hosts' do
  apache_dir = '/etc/apache2'

  %w(deflate expires headers rewrite).each do |mod|
    describe file("#{apache_dir}/mods-enabled/#{mod}.load") do
      it { should be_linked_to("../mods-available/#{mod}.load") }
    end
  end

  describe file("#{apache_dir}/ports.conf") do
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode '644' }
    its(:content) { should match(/Listen \*:8080\nNameVirtualHost \*:8080/) }
  end

  describe file("#{apache_dir}/conf.d/h5bp.conf") do
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode '644' }
  end

  %w(console web app).each do |app|
    docroot = "/var/www/#{app}.evertrue.com"

    describe file(docroot) do
      it { should be_directory }
      it { should be_owned_by 'deploy' }
      it { should be_grouped_into 'www-data' }
      it { should be_mode '2775' }
    end

    describe file("#{apache_dir}/sites-enabled/#{app}.conf") do
      it { should be_linked_to("../sites-available/#{app}.conf") }
    end

    describe file("#{apache_dir}/sites-available/#{app}.conf") do
      it { should be_owned_by 'root' }
      it { should be_grouped_into 'root' }
      it { should be_mode '644' }
      its(:content) { should include "ServerName local-#{app}.evertrue.com" }
      its(:content) { should include "DocumentRoot #{docroot}" }
    end
  end
end

describe 'Supporting functionality for Console/Web' do
  describe package('git') do
    it { should be_installed }
  end

  %w(node npm bower grunt).each do |bin|
    describe command("which #{bin}") do
      its(:stdout) { should include bin }
    end
  end
end


recipes/default.rb
#
# Cookbook Name:: et_console_app
# Recipe:: default
#
# Copyright (C) 2014 EverTrue, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

case node['platform_family']
when 'debian'
  include_recipe 'apt'
  package 'fontconfig'
end

package 'git'

include_recipe 'node'
include_recipe 'et_users::evertrue'
include_recipe 'apache2'

[
  node['et_console_app']['deploy_to'],
  node['et_web_app']['deploy_to'],
  node['et_app_app']['deploy_to']
].each do |deploy_to_dir|
  directory deploy_to_dir do
    owner 'deploy'
    group node['apache']['group']
    mode '2775'
  end
end

# Install some excellent Apache config rules, courtesy of h5bp.com
cookbook_file "#{node['apache']['dir']}/conf.d/h5bp.conf" do
  source 'h5bp.conf'
  mode '0644'
  owner 'root'
  group node['apache']['root_group']
  action :create_if_missing
end

web_app 'console' do
  server_name node['et_console_app']['server_name']
  server_aliases node['et_console_app']['server_aliases']
  docroot node['et_console_app']['docroot']
end

web_app 'web' do
  server_name node['et_web_app']['server_name']
  docroot node['et_web_app']['docroot']
end

web_app 'app' do
  server_name node['et_app_app']['server_name']
  docroot node['et_app_app']['docroot']
end
